python/pyarrow/tests/test_extension_type.py

```diff
@@ -1079,3 +1082,282 @@ def test_array_constructor_from_pandas():
+    >>> tensor_type
+    FixedShapeTensorType(FixedSizeListType(fixed_size_list<item: int32>[4]))
+    >>> pa.register_extension_type(tensor_type)
+    """
+
+    def __init__(self, value_type, shape, dim_names=None, permutation=None):
+        self._value_type = value_type
+        self._shape = shape
+        size = math.prod(shape)
+        self._dim_names = dim_names
+        self._permutation = permutation
+        pa.ExtensionType.__init__(self, pa.list_(self._value_type, size),
+                                  'arrow.fixed_size_tensor')
+
+    @property
+    def value_type(self):
+        """
+        Data type of an individual tensor.
+        """
+        return self._value_type
+
+    @property
+    def shape(self):
+        """
+        Shape of the tensors.
+        """
+        return self._shape
+
+    @property
+    def dim_names(self):
+        """
+        Explicit names of the dimensions.
+        """
+        return self._dim_names
+
+    @property
+    def permutation(self):
+        """
+        Indices of the dimensions ordering.
+        """
+        return self._permutation
+
+    def __arrow_ext_serialize__(self):
+        metadata = {"shape": str(self._shape),
+                    "dim_names": str(self._dim_names),
+                    "permutation": str(self._permutation)}
+        return json.dumps(metadata).encode()
+
+    @classmethod
+    def __arrow_ext_deserialize__(cls, storage_type, serialized):
+        # return an instance of this subclass given the serialized
+        # metadata.
+        assert serialized.decode().startswith('{"shape":')
+
+        metadata = json.loads(serialized.decode())
+        shape = ast.literal_eval(metadata['shape'])
+        dim_names = ast.literal_eval(metadata['dim_names'])
+        permutation = ast.literal_eval(metadata['permutation'])
+
+        return FixedShapeTensorType(storage_type.value_type, shape,
+                                    dim_names, permutation)
+
+    def __arrow_ext_class__(self):
+        return FixedShapeTensorArray
+
+
+class FixedShapeTensorArray(pa.ExtensionArray):
+    """
+    Canonical extension array class for fixed shape tensors.
+
+    Examples
+    --------
+    Define and register extension type for tensor array
+
+    >>> import pyarrow as pa
+    >>> tensor_type = FixedShapeTensorType(pa.int32(), (2, 2))
+    >>> pa.register_extension_type(tensor_type)
+
+    Create an extension array
+
+    >>> arr = [[1, 2, 3, 4], [10, 20, 30, 40], [100, 200, 300, 400]]
+    >>> storage = pa.array(arr, pa.list_(pa.int32(), 4))
+    >>> pa.ExtensionArray.from_storage(tensor_type, storage)
+    <__main__.FixedShapeTensorArray object at ...>
+    [
+      [
+        1,
+        2,
+        3,
+        4
+      ],
+      [
+        10,
+        20,
+        30,
+        40
+      ],
+      [
+        100,
+        200,
+        300,
+        400
+      ]
+    ]
+    """
+
+    def to_numpy_tensor(self):
```

Comment:
@jorisvandenbossche I haven't used permutations parameter in `to_numpy_tensor()` or in `from_numpy_tensor()` as the shape of the ndarray changes if the dimensions are permutated with [transpose](https://numpy.org/doc/stable/reference/generated/numpy.transpose.html). What do you think?

Reply:
I think that is fine to leave that out. It's up the user / application to know if they want to use the permutations anyway. You could maybe add an example of doing this using the available API: `np_arr = arr.to_numpy_tensor(); np_arr.transpose(arr.type.permutation)`

Reply:
I'd perhaps name these methods `to_numpy` and `from_numpy`. Tensor part is already implied from context and `Tensor` class [uses the same names](https://github.com/apache/arrow/blob/master/python/pyarrow/tests/test_tensor.py#L78).

Reply:
I wouldn't use `to_numpy()`, because that is an existing method on pyarrow Arrays, and from which you expect to get a 1D numpy array of the same length as the pyarrow array. 
While a nD array mostly will behave consistently as a 1D array (eg doing `result[0]` in both cases would give the same), I am still a bit hesitant to break the existing contract of `to_numpy()` in this case.

Reply:
That's fair. How about `to_numpy_array`? I'm skeptical of the word "tensor" as it doesn't appear in numpy.

Reply:
`to_ndarray`?

Reply:
I would personally keep "numpy" in the name, given that it's explicitly a numpy ndarray, and not from some other package

Reply:
`to_numpy_ndarray`?

Reply:
That's quite close to the original `to_numpy_tensor` ;) But either way, both are fine for me

Reply:
A nit: I think there are two schools of thought regarding naming tensors. One calls them tensors, the other multidimensional arrays. I think numpy is in the former camp and hence I'd avoid calling it `*_tensor`. :)

Reply:
@rok, can we leave as is or would you prefer ndarray to tensor?

Reply:
Sorry, GitHub failed to show me your last message while I had the site open. I will change to `to_numpy_ndarray` then.